cases/src/org/commcare/cases/util/CasePurgeFilter.java

```diff
@@ -222,7 +227,96 @@ private void propogateMarkToDAG(DAG<String, int[], String> dag, boolean walkFrom
         }
     }
 
-    private boolean caseStatusIs(int status, int flag) {
+    /**
+     * Traverse the graph to remove any edges to empty nodes (which are created when a child
+     * makes a placeholder index to a parent, but then the parent does not actually exist on the
+     * phone for some reason). Then remove any nodes that are made invalid by that parent node
+     * not existing
+     *
+     * @return Whether or not this method invocation removed any invalid edges from the DAG
+     */
+    private boolean removeInvalidEdges() {
+        Hashtable<String, Vector<Edge<String, String>>> allEdges = internalCaseDAG.getEdges();
+        Set<String> childOfNonexistentParent = new HashSet<>();
+        Set<String[]> edgesToRemove = new HashSet<>();
+        boolean invalidEdgesRemoved = false;
+        for (String originIndex : allEdges.keySet()) {
+            Vector<Edge<String, String>> edgeListForOrigin = allEdges.get(originIndex);
+            for (Edge<String, String> edge : edgeListForOrigin) {
+                String targetIndex = edge.i;
+                if (internalCaseDAG.getNode(targetIndex) == null) {
+                    invalidEdgesRemoved = true;
+                    edgesToRemove.add(new String[]{originIndex, targetIndex});
+                    childOfNonexistentParent.add(originIndex);
+                }
+            }
+        }
+
+        // Any case node with a nonexistent parent should be removed from the graph, as should
+        // all of its descendants, and any edges to or from it
+        for (String index : childOfNonexistentParent) {
+            edgesToRemove = removeNodeAndPropagate(index, edgesToRemove);
```

Comment:
I'm a big confused, this both takes in edgesToRemove as an assignment but also passes it in as an argument. It's an accumulator, right? If so we shouldn't need to be passing it back.